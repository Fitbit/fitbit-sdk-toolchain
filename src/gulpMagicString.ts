import { Transform } from 'stream';

import MagicString from 'magic-string';
import PluginError from 'plugin-error';
import {
  SourceMapConsumer,
  SourceMapGenerator,
  RawSourceMap,
} from 'source-map';
import Vinyl from 'vinyl';

const PLUGIN_NAME = 'gulp-magic-string';

async function mergeSourceMaps(inMap: RawSourceMap, outMap: RawSourceMap) {
  return SourceMapConsumer.with(inMap, null, async (inConsumer) =>
    SourceMapConsumer.with(outMap, null, async (outConsumer) => {
      const generator = SourceMapGenerator.fromSourceMap(inConsumer);
      generator.applySourceMap(outConsumer);
      return generator.toJSON();
    }),
  );
}

export default function gulpMagicString(
  transformerFunc: (code: string, magicString: MagicString) => void,
) {
  const transform = new Transform({
    objectMode: true,
    transform(file: Vinyl, _, next) {
      if (file.isNull() || file.extname !== '.js') {
        return next(undefined, file);
      }

      if (!file.isBuffer()) {
        next(
          new PluginError(
            PLUGIN_NAME,
            file.isStream()
              ? 'Streaming mode is not supported.'
              : 'Internal error processing file.',
            { fileName: file.relative },
          ),
        );
        return;
      }

      try {
        const code = file.contents.toString('utf-8');
        const parsed = new MagicString(code);

        transformerFunc(code, parsed);

        file.contents = Buffer.from(parsed.toString());

        if (!file.sourceMap) return next(undefined, file);
        const map = parsed.generateMap();

        mergeSourceMaps(
          file.sourceMap,
          // magic-string thinks map.version is a string, source-map thinks it's a number
          {
            ...map,
            version: Number(map.version),
            file: file.relative,
            sources: [file.relative],
          },
        ).then((sourceMap) => {
          file.sourceMap = sourceMap;
          next(undefined, file);
        });
      } catch (error) {
        next(new PluginError(PLUGIN_NAME, error, { fileName: file.relative }));
      }
    },
  });

  return transform;
}
