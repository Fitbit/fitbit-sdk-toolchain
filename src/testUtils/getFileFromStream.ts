import stream from 'stream';

import Vinyl from 'vinyl';

export default function getFileFromStream(stream: stream.Readable, filename?: string): Promise<Vinyl> {
    return new Promise<Vinyl>((resolve, reject) => {
        stream.on('data', (file: Vinyl) => {
            if (file.contents && (!filename || file.basename === filename)) {
                resolve(file);
            }
        });
        stream.on('finish', () => reject(new Error('No file in stream matched')));
        stream.on('error', reject);
    });
}
