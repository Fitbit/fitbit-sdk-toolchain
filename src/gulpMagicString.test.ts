import Vinyl from 'vinyl';

import gulpMagicString from './gulpMagicString';
import getFileFromStream from './testUtils/getFileFromStream';

it('emits a valid sourcemap', async () => {
    const stream = gulpMagicString((code, magic) => magic.appendRight(code.length, ';'));

    const inputCode = 'console.log("hi!");\n';
    const inputPath = 'index.js';

    stream.write(
        new Vinyl({
            path: inputPath,
            contents: Buffer.from(inputCode),
            sourceMap: {
                version: 3,
                sources: [inputPath],
                names: ['console', 'log'],
                mappings: ';AAAAA,QAAAC,IAAY',
                file: inputPath,
                sourcesContent: [inputCode],
            },
        })
    );

    const { sourceMap } = await getFileFromStream(stream, inputPath);
    expect(sourceMap).toMatchSnapshot();
});
