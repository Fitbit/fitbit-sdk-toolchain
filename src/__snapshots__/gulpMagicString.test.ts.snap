// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits a valid sourcemap 1`] = `
Object {
  "file": "index.js",
  "mappings": ";AAAAA,QAAAC,IAAA",
  "names": Array [
    "console",
    "log",
  ],
  "sources": Array [
    "index.js",
  ],
  "sourcesContent": Array [
    "console.log(\\"hi!\\");
",
  ],
  "version": 3,
}
`;
